src/main/java/com/appland/appmap/record/Recorder.java

```diff
@@ -22,202 +18,189 @@
-   * the queue.
-   */
-  private void flush(IRecordingSession recordingSession) throws ActiveSessionException {
-    Collection<Event> events;
-
-    synchronized (this) {
-      events = new LinkedList<Event>(this.queuedEvents.values());
-      this.queuedEvents.clear();
-    }
-
-    for (Event event : events) {
-      this.writeEvent(event, recordingSession);
-    }
-  }
-
-  private void queueEvent(Event event) throws ActiveSessionException {
-    Event pendingEvent;
-    IRecordingSession recordingSession;
-
-    synchronized (this) {
-      if (this.activeSession == null) {
-        return;
-      }
-
-      recordingSession = this.activeSession;
-      pendingEvent = this.queuedEvents.get(event.threadId);
-      this.queuedEvents.put(event.threadId, event);
-    }
-
-    if (pendingEvent != null) {
-      this.writeEvent(pendingEvent, recordingSession);
-    }
-  }
-
-  private synchronized void forceStop() {
-    if (this.activeSession == null) {
-      return;
-    }
-
-    this.activeSession.stop();
-    this.activeSession = null;
-  }
-
-  /**
-   * Stops the active recording session.
-   * @return Output from the current session. This will be empty unless recording to memory.
+   * Stops the active recording session and obtains the result.
+   *
+   * @return Recording of the current session.
    * @throws ActiveSessionException If no recording session is in progress or the session cannot be
    *                                stopped.
    */
-  public String stop() throws ActiveSessionException {
-    IRecordingSession recordingSession;
-
-    synchronized (this) {
-      recordingSession = this.getActiveSession();
-      this.activeSession = null;
-    }
-
-    try {
-      this.flush(recordingSession);
-      return recordingSession.stop();
-    } catch (ActiveSessionException e) {
-      Logger.printf("failed to stop recording\n%s\n", e.getMessage());
-      this.forceStop();
-      return "";
-    }
+  public Recording stop() throws ActiveSessionException {
+    this.flush();
+    return activeSession.release().stop();
   }
 
   /**
    * Record an {@link Event} to the active session.
+   *
    * @param event The event to be recorded.
    */
   public void add(Event event) {
+    if (!activeSession.exists()) {
+      return;
+    }
+
+    ThreadState ts = threadState();
+
+    // We don't want re-entrant events on the same thread.
+    if ( ts.isProcessing ) {
+      return;
+    }
+
+    ts.isProcessing = true;
     try {
-      this.queueEvent(event);
-    } catch (ActiveSessionException e) {
-      Logger.println("failed to record event");
-      Logger.println(e);
-      this.forceStop();
+      if ( event.event.equals("call") ) {
+        ts.callStack.push(event);
+      } else if ( event.event.equals("return") ) {
+        if ( ts.callStack.isEmpty() ) {
+          Logger.println("Discarding 'return' event because the call stack is empty for this thread");
```

Comment:
This could happen if a recording is started and the first event that's recorded is a `return`.